### PR TITLE
preference file implementation

### PR DESCRIPTION
**Pull request by surajssd**
now user can provider a prefernce file where s(he) can mention what controllers to generate and what provider to default to.

### CODE REVIEW

**Comment by surajssd** [comment]
using following docker-compose file
```yaml
$ cat docker-compose.yml 
version: "2"

services:
  mariadb:
    image: centos/mariadb
    ports:
      - "$DB_PORT"
    environment:
      MYSQL_ROOT_PASSWORD: $ROOT_PASS
      MYSQL_DATABASE: $DB_NAME
      MYSQL_PASSWORD: $DB_PASS
      MYSQL_USER: $DB_USER

  etherpad:
    image: centos/etherpad
    ports:
      - "80:9001"
    depends_on:
      - mariadb
    environment:
      DB_HOST: $DB_HOST
      DB_DBID: $DB_NAME
      DB_PASS: $DB_PASS
      DB_PORT: $DB_PORT
      DB_USER: $DB_USER
```
and following preference file, right now this is default name `kompose.yml`

```yaml
$ cat kompose.yml 
profiles:
  default:
    provider: kubernetes
    objects:
      - deployment
      - replicationcontroller

  test:
    provider: openshift
    objects:
      - deploymentconfig

current-profile: test
```

here it read from the kompose file that provider is `openshift` and the controller to be created is `deploymentconfig`
```bash
$ kompose convert 
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deploymentconfig.json" created 
INFO[0000] file "etherpad-imagestream.json" created     
INFO[0000] file "mariadb-deploymentconfig.json" created 
INFO[0000] file "mariadb-imagestream.json" created      
```

now if i give provider as `kubernetes` from commandline then the pref file gets over-ridden

```bash
$ kompose --provider kubernetes convert --rc --d
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-deployment.json" created       
INFO[0000] file "mariadb-replicationcontroller.json" created 
INFO[0000] file "etherpad-deployment.json" created      
INFO[0000] file "etherpad-replicationcontroller.json" created 
```

now i change the `current-profile` in `kompose.yml` to `default`


pref file looks like this now
```yaml
$ cat kompose.yml 
profiles:
  default:
    provider: kubernetes
    objects:
      - deployment
      - replicationcontroller

  test:
    provider: openshift
    objects:
      - deploymentconfig

current-profile: default
```

normal conversion where data is read from `kompose.yml`, it creates `deployment` and `replicationcontroller` since it is mentioned in file
```bash
$ kompose convert
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deployment.json" created      
INFO[0000] file "etherpad-replicationcontroller.json" created 
INFO[0000] file "mariadb-deployment.json" created       
INFO[0000] file "mariadb-replicationcontroller.json" created 
```

i can also provide from commandline what all things i would want more, like here i ask to create `daemonset` as well using flag `--ds`

```bash
$ kompose convert --ds                                                                                                       
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deployment.json" created      
INFO[0000] file "etherpad-daemonset.json" created       
INFO[0000] file "etherpad-replicationcontroller.json" created 
INFO[0000] file "mariadb-deployment.json" created       
INFO[0000] file "mariadb-daemonset.json" created        
INFO[0000] file "mariadb-replicationcontroller.json" created 
```

i can always override from commandline
```bash
$ kompose --provider openshift convert
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deploymentconfig.json" created 
INFO[0000] file "etherpad-imagestream.json" created     
INFO[0000] file "mariadb-deploymentconfig.json" created 
INFO[0000] file "mariadb-imagestream.json" created   
```

**Comment by surajssd** [comment]
I will add tests and fix existing ones once I get a review that I am doing the right thing!

**Comment by sebgoa** [comment]
Yes that looks like the behavior we want.

Make sure it works even if the kompose.yml file is not there.

Then if we can read a local kompose.yml it takes precedence.

**Comment by surajssd** [comment]
if kompose.yml file not provided, defaults to kubernetes as before
```bash
$ kompose convert
WARN[0000] Error Unmarshalling file - kompose.yml: open kompose.yml: no such file or directory 
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deployment.json" created      
INFO[0000] file "mariadb-deployment.json" created       
```

provider if mentioned
```bash
$ kompose --provider openshift convert                                                                                       
WARN[0000] Error Unmarshalling file - kompose.yml: open kompose.yml: no such file or directory 
WARN[0000] Unsupported key depends_on - ignoring        
INFO[0000] file "etherpad-service.json" created         
INFO[0000] file "mariadb-service.json" created          
INFO[0000] file "etherpad-deploymentconfig.json" created 
INFO[0000] file "etherpad-imagestream.json" created     
INFO[0000] file "mariadb-deploymentconfig.json" created 
INFO[0000] file "mariadb-imagestream.json" created      
```
so if there is any kind of error while reading kompose.yml, rather than exiting it continues with defaults further.

**Comment by dustymabe** [comment]
> WARN[0000] Error Unmarshalling file - kompose.yml: open kompose.yml: no such file or directory 

This should probably check if the file exists and giving a friendlier warning message than this or even no warning at all (maybe info instead) because it is a perfectly normal use case to not have a preference file

File: pkg/config/validate.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/xeipuuv/gojsonschema"
+)
+
+func Validate(prefFile string) (kobject.ConvertOptions, error) {
+	var opt kobject.ConvertOptions
+
+	// Get the preference file name
+	prefFileContents, err := ioutil.ReadFile(prefFile)
+	if err != nil {
+		return opt, fmt.Errorf("Error Unmarshalling file - %s: %v", prefFile, err)
```
Comment (pradeepto) [inline pkg/config/validate.go]: @surajssd minor nitpick, this is not unmarshalling error, I think. Error message needs fixing.

**Comment by surajssd** [comment]
@dustymabe @pradeepto addressed your concerns, also added tests!

**Comment by kadel** [comment]
I don't think that Kompose should display warning if user didn't specified `--pref-file`. (default kompose.yml is used)
On the other hand it should be error if user explicitly set `--perf-file` to non-existing file.

**Comment by surajssd** [comment]
@kadel the problem here is that i have `pref-file` with default name, so there is no clean way to know if user has explicitly provided a value of not.

**Comment by kadel** [comment]
So one option would be to check if `GlobalString("pref-file") == DefaultPreferenceFile` then ignore warning. But this is not going to work if user sets `pref-file=kompose.yml` :-).

Or we can remove default value from flag definition and then you can check if user set it or not.

**Comment by surajssd** [comment]
> Or we can remove default value from flag definition and then you can check if user set it or not.

But then this makes it mandatory for user to provide file with flag!

**Comment by kadel** [comment]
> But then this makes it mandatory for user to provide file with flag!

Does it? I don't think so. If you remove `Value: app.DefaultPreferenceFile` from `CommonFlags`
than `pref-file` will be empty when user don't provide flag.

**Comment by surajssd** [comment]
@kadel what i mean by that is I cannot implicitly read a pref-file, user has to provide a file name/path with help of flag. So if you want to use pref file, it becomes mandatory to provide a file path for it.

**Comment by kadel** [comment]
I'm lost :-(

This is how i think it should behave:

user runs `kompose convert` and `kompose.yml` doesn't exist - no warning no error everything is fine

user runs `kompose convert` and `kompose.yml` exists - no warning no error everything is fine, and `kopmose.yml` is used

user runs `kompose --pref-file=mykompose.yml` and `mykompose.yml` exists - no warning no error everything is fine and `mykompose.yml` is used

user runs `kompose --pref-file=mykompose.yml` and `mykompose.yml` desn't exists - show error and fail.

**Comment by surajssd** [comment]
@kadel changes done, as per your comments!

**Comment by surajssd** [comment]
@sebgoa @ngtuna rebased please give it a try once and let me know what is missing :)

**Comment by surajssd** [comment]
Updated with unit-tests.

**Comment by surajssd** [comment]
closing in favour ot https://github.com/kubernetes-incubator/kompose/pull/366

**Pull request closed without merging**